---
title: "Weinstock Pomp Micobiome Samples"
author: "Duy Pham"
date: "1/4/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



```{r Options and Libraries, echo=FALSE, message=FALSE, warning=FALSE}
library(knitr)
library(tidyverse)
library(openxlsx)
```







```{r Read in Data, echo=FALSE}
sample_sheet <- read.xlsx("~/Desktop/Weinstock Microbiome/16S_taxonomic_abundance_tables_for_Pomp_DO_mouse_data/taxonomic_abundance_filtered_unnormalized_20181004.xlsx", sheet = 'Sample IDs')
raw_otu      <- read.xlsx("~/Desktop/Weinstock Microbiome/16S_taxonomic_abundance_tables_for_Pomp_DO_mouse_data/taxonomic_abundance_filtered_unnormalized_20181004.xlsx", sheet = 'otu_abundance')
raw_genus    <- read.xlsx("~/Desktop/Weinstock Microbiome/16S_taxonomic_abundance_tables_for_Pomp_DO_mouse_data/taxonomic_abundance_filtered_unnormalized_20181004.xlsx", sheet = 'genus_abundance')
raw_family   <- read.xlsx("~/Desktop/Weinstock Microbiome/16S_taxonomic_abundance_tables_for_Pomp_DO_mouse_data/taxonomic_abundance_filtered_unnormalized_20181004.xlsx", sheet = 'family_abundance')
raw_order    <- read.xlsx("~/Desktop/Weinstock Microbiome/16S_taxonomic_abundance_tables_for_Pomp_DO_mouse_data/taxonomic_abundance_filtered_unnormalized_20181004.xlsx", sheet = 'order_abundance')
raw_class    <- read.xlsx("~/Desktop/Weinstock Microbiome/16S_taxonomic_abundance_tables_for_Pomp_DO_mouse_data/taxonomic_abundance_filtered_unnormalized_20181004.xlsx", sheet = 'class_abundance')
raw_phylum   <- read.xlsx("~/Desktop/Weinstock Microbiome/16S_taxonomic_abundance_tables_for_Pomp_DO_mouse_data/taxonomic_abundance_filtered_unnormalized_20181004.xlsx", sheet = 'phylum_abundance')


genotype <- readRDS("~/Desktop/Weinstock Microbiome/Original Genotypes/pomp_genoprobs_qtl2.rds")
```




```{r Subset samples by week, echo=FALSE}

### Get sampels with genotypes
genotype_samples <- dimnames(genotype[[1]])[[1]]
genotype_samples <- grep('DO2', genotype_samples, value = TRUE)






### Get samples for each week
sample_sheet <- sample_sheet %>%
                             arrange(SampleName) %>%
                             unique() 
original_w6_counts <- sample_sheet %>%
                                   filter(grepl('w6', SampleName))
original_w17_counts <- sample_sheet %>%
                                   filter(grepl('w17', SampleName))
original_w24_counts <- sample_sheet %>%
                                   filter(grepl('w24', SampleName))



w6_names <- paste0(paste0('DPDP.DO2.',original_w6_counts$Mouse), '.F')
w17_names <- paste0(paste0('DPDP.DO2.',original_w17_counts$Mouse), '.F')
w24_names <- paste0(paste0('DPDP.DO2.',original_w24_counts$Mouse), '.F')


```





```{r Get Sample Counts, echo=FALSE}

### Get original number of samples in data
overlap_originals <- Reduce(intersect, list(genotype_samples, w6_names, w17_names, w24_names))

original_sample_counts <- c(paste(length(genotype_samples), '(2 dups)'), 
                            nrow(original_w6_counts), 
                            nrow(original_w17_counts), 
                            nrow(original_w24_counts),
                            length(overlap_originals))






### Get samples with genotypes and count data
samples_with_counts_and_genotypes_w6  <- intersect(genotype_samples, paste0(paste0('DPDP.DO2.', original_w6_counts$Mouse),'.F'))
samples_with_counts_and_genotypes_w17 <- intersect(genotype_samples,paste0(paste0('DPDP.DO2.', original_w17_counts$Mouse),'.F'))
samples_with_counts_and_genotypes_w24 <- intersect(genotype_samples,paste0(paste0('DPDP.DO2.', original_w24_counts$Mouse),'.F'))


samples_with_counts_and_genotypes_overlap <- Reduce(intersect, list(genotype_samples, 
                                                                    samples_with_counts_and_genotypes_w6,
                                                                    samples_with_counts_and_genotypes_w17,
                                                                    samples_with_counts_and_genotypes_w24))


samples_with_genotype_and_microbiome_counts <- c(NA,
                                                 length(samples_with_counts_and_genotypes_w6),
                                                 length(samples_with_counts_and_genotypes_w17),
                                                 length(samples_with_counts_and_genotypes_w24),
                                                 length(samples_with_counts_and_genotypes_overlap))

```

```{r Mixture Samples, echo=FALSE}
week_6_mixtures  <- paste0(paste0('DPDP.', c('DO2.444','DO2.492','DO2.507','DO2.541','DO2.552')), '.F')
week_17_mixtures <- paste0(paste0('DPDP.', c('DO2.440','DO2.484','DO2.833','DO2.835','DO2.836','DO2.840','DO2.841','DO2.850')), '.F')
week_24_mixtures <- paste0(paste0('DPDP.', c('DO2.411','DO2.416','DO2.472','DO2.504','DO2.539','DO2.549','DO2.732','DO2.801','DO2.825','DO2.846')), '.F')



samples_with_counts_and_genotypes_and_nomixtures_w6  <- samples_with_counts_and_genotypes_w6[-match(week_6_mixtures, samples_with_counts_and_genotypes_w6)]
samples_with_counts_and_genotypes_and_nomixtures_w17 <- samples_with_counts_and_genotypes_w17[-match(week_17_mixtures, samples_with_counts_and_genotypes_w17)]
samples_with_counts_and_genotypes_and_nomixtures_w24 <- samples_with_counts_and_genotypes_w24[-match(week_24_mixtures, samples_with_counts_and_genotypes_w24)]
samples_with_counts_and_genotypes_and_nomixtures_overlap <- Reduce(intersect, list(samples_with_counts_and_genotypes_and_nomixtures_w6,
                                                                                   samples_with_counts_and_genotypes_and_nomixtures_w17,
                                                                                   samples_with_counts_and_genotypes_and_nomixtures_w24))


samples_with_genotype_and_microbiome_counts_and_no_mixture <- c(NA,
                                                                length(samples_with_counts_and_genotypes_and_nomixtures_w6),
                                                                length(samples_with_counts_and_genotypes_and_nomixtures_w17),
                                                                length(samples_with_counts_and_genotypes_and_nomixtures_w24),
                                                                length(samples_with_counts_and_genotypes_and_nomixtures_overlap))

```



```{r Combine Results, echo=FALSE}

counts_df <- as.data.frame(rbind(original_sample_counts,
                                 samples_with_genotype_and_microbiome_counts,
                                 samples_with_genotype_and_microbiome_counts_and_no_mixture))


colnames(counts_df) <- c('Genotype','Week 6', 'Week 17', 'Week 24', 'Overlap')
rownames(counts_df) <- c('Original Sample Counts',
                         'Samples with Genotype Prob and Microbiome Counts',
                         'Samples with Genotype Prob and Microbiome Counts and No Mixture')


kable(counts_df, caption = 'Number of Samples with Genotype Probabilities and Microbiome Count')
```




